src/MongoDB.Driver.Core/Core/Bindings/CoreSession.cs

```diff
@@ -434,17 +435,41 @@ private void EnsureStartTransactionCanBeCalled()
         {
             if (_currentTransaction == null)
             {
-                return;
+                EnsureTransactionsAreSupported();
             }
+            else
+            {
+                switch (_currentTransaction.State)
+                {
+                    case CoreTransactionState.Aborted:
+                    case CoreTransactionState.Committed:
+                        break;
 
-            switch (_currentTransaction.State)
+                    default:
+                        throw new InvalidOperationException("Transaction already in progress.");
+                }
+            }
+        }
+
+        private void EnsureTransactionsAreSupported()
+        {
+            var connectedDataBearingServers = _cluster.Description.Servers.Where(s => s.State == ServerState.Connected && s.IsDataBearing).ToList();
+
+            if (connectedDataBearingServers.Count == 0)
             {
-                case CoreTransactionState.Aborted:
-                case CoreTransactionState.Committed:
-                    return;
+                throw new NotSupportedException("StartTransaction cannot determine if transactions are supported because there are no connected servers.");
+            }
 
-                default:
-                    throw new InvalidOperationException("Transaction already in progress.");
+            foreach (var connectedDataBearingServer in connectedDataBearingServers)
+            {
+                if (connectedDataBearingServer.Type == ServerType.ShardRouter)
+                {
+                    Feature.ShardedTransactions.ThrowIfNotSupported(connectedDataBearingServer.Version);
```

Comment:
Is there a test for this?  I wasn't able to find one.

Reply:
See:

    
 
 
 
 EnsureTransactionsAreSupported_should_throw_when_any_connected_replica_set_member_does_not_support_transactions

Which I will rename because after some refactoring it testing is not limited to replica set members.